spring-integration-core/src/test/java/org/springframework/integration/handler/MethodInvokingMessageProcessorTests.java

```diff
@@ -348,6 +349,110 @@ public void multipleHeadersWithAnnotatedMethod() throws Exception {
 	}
 
 	@Test
+	public void optionalAndRequiredWithAnnotatedMethod() throws Exception {
+		AnnotatedTestService service = new AnnotatedTestService();
+		Method method = service.getClass().getMethod("optionalAndRequiredHeader", String.class, Integer.class);
+		MethodInvokingMessageProcessor processor = new MethodInvokingMessageProcessor(service, method);
+		processor.setUseSpelInvoker(true);
+		optionalAndRequiredWithAnnotatedMethodGuts(processor, false);
+	}
+
+	@Test
+	public void compiledOptionalAndRequiredWithAnnotatedMethod() throws Exception {
+		AnnotatedTestService service = new AnnotatedTestService();
+		Method method = service.getClass().getMethod("optionalAndRequiredHeader", String.class, Integer.class);
+		MethodInvokingMessageProcessor processor = new MethodInvokingMessageProcessor(service, method);
+		processor.setUseSpelInvoker(true);
+		DirectFieldAccessor compilerConfigAccessor = compileImmediate(processor);
+		optionalAndRequiredWithAnnotatedMethodGuts(processor, true);
+		assertNotNull(TestUtils.getPropertyValue(processor, "delegate.handlerMethod.expression.compiledAst"));
+		optionalAndRequiredWithAnnotatedMethodGuts(processor, true);
```

Comment:
OK. Now I see complex logic in the target `Guts` method.

Thanks.
So, merging...